### PR TITLE
 Fix #168: avoid creating duplicate link for addon 

### LINKED ISSUE #168: Blender: Start creates copy of add-on when running in Blender 4.2
Running Blender:Start will make a copy of the add-on, meaning two add-on are run at the same time. The orginal addon is inside blender_org, a copy will be added to user_default. If the add-on is in the user_default from the start Blender crashes, so its seems it only works in blender_org, but then a copy is created.

image
image

### PR DESCRIPTION
**Pull request by Mateusz-Grzelinski**
In this PR:

- allow user to develop addon even it is placed in directories like (without creating duplicate addon link):
    - `\4.2\scripts\addons` -> default dir for addons
    - `\4.2\extensions\blender_org` -> directory indicated by `bpy.context.preferences.extensions.repos` (list of directories) 
- use python3.7 syntax for compability
- ensured (basic) compability with blender 2.80
- remove duplicate links to development (VSCode) directory
- remove broken links in addon and extension dir

Squash merge when ready. Please squash merge after https://github.com/JacquesLucke/blender_vscode/pull/174 - it fixes one more issue with 2.80 compability

About tests: 
- they are mostly for me, just nice to have. if you look closesly they are missing many asserts and many errors will slip through when tesing on real blender instance.
- should be cross platform, but tested only on windows
- the tests run outside of blender. `bpy` and `addon_utils` are not necessary for tests
- how to run:
```powershell
pip install pytest
cd pythonFile
$env:PYTHONPATH="./include" # powershell
pytest -s .\tests 
```

I could not figure out how to run tests with VS code UI. Works fine with pycharm UI when you mark include directory as source directory.

todo:
- [x] fix tests

Corner cases (for documentation):
- junctions and hard links are difficult to manage
  - python behaves completely differently when touching junction in different python versions!
- link to extension is already present in different repository - remove the link
- link to addon is already present in addon directory but has different name - remove the link
- You are developing extension in addons directory (legacy addon not supported) - make a link
- You are developing addon in extensions directory (no extension support) - make a link
  - why would you do that to me...
- extension has legacy bl_info defined (supportes both legacy and new addon systems) - prefer extension for >4.2, otherwise addon
- extension has legacy bl_info defined (supportes both legacy and new addon systems) BUT user is developing addon in addon directory - treat as addon - do not make a link.

### CODE REVIEW

**Comment by Mateusz-Grzelinski** [comment]
The change should be working now.
I tried adding some basic tests, but I did something wrong, because 
- if you run tests one by one they work
- if you run the whole file the tests fail - the patch bleeds into other functions. I have no idea how to fix it.

Note: I have basic experience with tests.

```powershell
pip install pytest
cd pythonFile
$env:PYTHONPATH="./include" # powershell
pytest -s .\tests\blender_vscode\test_load_addons.py::TestIsInAnyAddonDirectory::test_is_in_any_addon_directory # ok
pytest -s .\tests  # 1 ok 5 fail
```

**Comment by Mateusz-Grzelinski** [comment]
I think I fixed the tests, but i need to sleep on it.
- rooky mistake: patch calls where they are made
- confusing patching behaviour: you can not patch the attributes in same path in two places (probably)
- removing modules works differently than I anticipated
- mocking global variables is hard
- fixed python 3.7 compability
Those tests are just prototypes that helped me find 2 more errors in my code. I am not sure if this is the right way to test

File: pythonFiles/include/blender_vscode/load_addons.py [hunk 1]
```diff
@@ -11,25 +12,30 @@
...
-            load_path = addon_info.load_dir
+        default_directory = get_user_addon_directory(Path(addon_info.load_dir))
+        if is_addon_legacy(Path(addon_info.load_dir)):
+            if is_in_any_addon_directory(addon_info.load_dir):
+                # blender knows about addon and can load it
+                load_path = addon_info.load_dir
+            else:  # is in external dir of is in extensions dir
+                load_path = os.path.join(default_directory, addon_info.module_name)
+                if str(load_path) not in sys.path:
+                    sys.path.append(str(load_path))
```
Comment (Mateusz-Grzelinski) [inline pythonFiles/include/blender_vscode/load_addons.py:27]: I am not sure when adding path is required

**Comment by Mateusz-Grzelinski** [comment]
* found even more issues (and fixed them)
* made a super-duper long test. The test is not that complicated but the patching is complicated.

**Comment by Mateusz-Grzelinski** [comment]
I am pretty sure this patch works as expected. Final test was done by adding addon and extension to both
- `\4.2\scripts\addons`
- `\4.2\extensions\blender_org` 

as seen on screeshot, all looks ok.
image

The policy violation was fixed in https://github.com/JacquesLucke/blender_vscode/pull/172/commits/445778f950fcc2e33bd62310e98a31d9c8bdc539

File: pythonFiles/include/blender_vscode/load_addons.py (hunk 1 above)
Comment (Mateusz-Grzelinski) [reply pythonFiles/include/blender_vscode/load_addons.py:27]: adding sys.path for extensions resylts in blneder warning about policy violation

File: pythonFiles/tests/blender_vscode/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+scripts
```
Comment (JacquesLucke) [inline pythonFiles/tests/blender_vscode/.gitignore:1]: Looks like this is missing a newline.

**Review by JacquesLucke — COMMENTED** [review]
Would be good if you could update the initial PR description to be a bit more comprehensive. 

A few words about how to run the tests would be good too. Also looks like they work on linux only (it's ok, but would be good to be more explicit about it)?

**Comment by Mateusz-Grzelinski** [comment]
ugh, breaks older blender. Totally forgot about that. I need to fight the persmission denied error and check with blender 2.8 - i think this is the oldest one we support.

**Comment by Mateusz-Grzelinski** [comment]
fixed compability with bledner 2.80. Please squash merge after #174 - it fixes one more issue with 2.80 compability

**Comment by Mateusz-Grzelinski** [comment]
updated to newest master, ready to merge

**Comment by Mateusz-Grzelinski** [comment]
Covered one more edge case 

1. You are developing extension in addons directory 
2. extension has legacy bl_info defined (supportes both legacy and new addon systems)
3. then do not create a link (it will be duplicate)

**Comment by Mateusz-Grzelinski** [comment]
the tests are fixed. but I just thought about yet another corner case:
- link to extension is already present in different repository - this will cause duplicate
- link to addon is already present in addon directory but has different name - also a duplicate

**Comment by Mateusz-Grzelinski** [comment]
there are problems with python 3.7 and dealing with junctions in windows. They are not recognized correctly and give different errors than python 3.12

File: pythonFiles/include/blender_vscode/load_addons.py [hunk 3]
```diff
@@ -1,67 +1,216 @@
...
+            load_path = _link_addon_or_extension(addon_info)
+        except PermissionError as e:
+            print(
+                f"""ERROR: {e} 
+Path "{e.filename}" can not be removed. **Please remove it manually!** Most likely causes:
+    - Path requires admin permissions to remove
+    - Windows only: You upgraded Blender version and imported old setting. Now links became real directories.
+    - Path is a real directory with the same name as addon (removing might cause data loss!)"""
+            )
+            raise e
```
Comment (Mateusz-Grzelinski) [inline pythonFiles/include/blender_vscode/load_addons.py:46]: technically, there is no need to "panic: here, we can still load other addons, but there is no indication that linking failed and bledner needs to be restarted anyways...

**Comment by Mateusz-Grzelinski** [comment]
tested on ubuntu 22.04 snap installed blender 4.2. Works fine. 
Merging.
